osquery/tables/system/posix/sudoers.cpp

```diff
@@ -27,42 +31,122 @@ const std::string kSudoFile = "/etc/sudoers";
+  if (level > kMaxNest) {
+    TLOG << "sudoers file recursion maximum reached";
+    return;
   }
 
   std::string contents;
-  if (!forensicReadFile(kSudoFile, contents).ok()) {
-    return results;
+  if (!forensicReadFile(filename, contents).ok()) {
+    TLOG << "couldn't read sudoers file: " << filename;
+    return;
   }
 
   auto lines = split(contents, "\n");
-  std::vector<std::string> valid_lines;
-
   for (auto& line : lines) {
+    Row r;
     boost::trim(line);
 
     // Only add lines that are not comments or blank.
     if (line.size() > 0 && line.at(0) != '#') {
-      valid_lines.push_back(line);
+      r["source"] = filename;
+
+      auto cols = split(line);
+      r["header"] = cols.at(0);
+
+      cols.erase(cols.begin());
+      r["rule_details"] = join(cols, " ");
+
+      results.push_back(r);
+    } else if (line.find("#includedir") == 0) {
+      auto space = line.find_first_of(' ');
+
+      // If #includedir doesn't look like it's followed by
+      // a path, treat it like a normal comment.
+      if (space == std::string::npos) {
+        continue;
+      }
+
+      auto inc_dir = line.substr(space + 1);
+
+      // NOTE(ww): See sudo NEWS for 1.8.4:
+      // Both #include and #includedir support relative paths.
+      if (inc_dir.at(0) != '/') {
+        auto path = fs::path(filename).parent_path() / inc_dir;
+        inc_dir = path.string();
+      }
+
+      // Build and push the row before recursing.
+      r["source"] = filename;
+      r["header"] = "includedir";
+      r["rule_details"] = inc_dir;
+      results.push_back(r);
+
+      std::vector<std::string> inc_files;
+      if (!listFilesInDirectory(inc_dir, inc_files).ok()) {
+        TLOG << "couldn't list includedir: " << inc_dir;
+        continue;
+      }
+
+      for (const auto& inc_file : inc_files) {
+        std::string inc_basename = fs::path(inc_file).filename().string();
+
+        // Per sudoers(5): Any files in the included directory that
+        // contain a '.' or end with '~' are ignored.
+        if (inc_basename.empty() ||
+            inc_basename.find('.') != std::string::npos ||
+            inc_basename.back() == '~') {
+          continue;
+        }
+
+        genSudoersFile(inc_file, ++level, results);
+      }
+    } else if (line.find("#include") == 0) {
+      auto space = line.find_first_of(' ');
+
+      // If #include doesn't look like it's followed by
+      // a path, treat it like a normal comment.
+      if (space == std::string::npos) {
+        continue;
+      }
+
+      auto inc_file = line.substr(space + 1);
+
+      // Per sudoers(5): If the included file doesn't
+      // start with /, read it relative to the current file.
+      if (inc_file.at(0) != '/') {
+        auto path = fs::path(filename).parent_path() / inc_file;
+        inc_file = path.string();
+      }
+
+      r["source"] = filename;
+      r["header"] = "include";
+      r["rule_details"] = inc_file;
+      results.push_back(r);
```

Comment:
Just for the sake of performance, you could move `r` into the push_back, to prevent useless copy.
```C++
results.push_back(std::move(r));
```